database/src/adapters/genesis.rs

```diff
@@ -0,0 +1,59 @@
+use actix_diesel::Database;
```

Comment:
The [`master-nearcore` version](https://github.com/near/near-indexer-for-explorer/blob/master-nearcore/src/db_adapters/genesis.rs#L22) of this file included code to 'execute an async body on a specified actix runtime'. I'm not sure if this is still necessary as the excluded in still seemed to work.

Reply:
While it still works doesn't mean "in still" works faster than async. Perhaps it is worth keeping.

@frol do you have any thoughts on that, please?

Reply:
@morgsmccauley It was essential back then because otherwise we halt all the async threads on the blocking read and blocking JSON parsing of genesis file (on testnet it is 5GB JSON file and takes 5-10 minutes to get through it). In the case of nearcore, we blocked p2p communication and other tasks, and then it did not feel healthy. Given that we don't do anything fancy in background, we can get away with blocking the whole async runtime here, but I would still consider extracting it into a separate thread just to avoid unnecessary surprises if `store_genesis_records` would be used under different circumstances assuming it would cooperate (won't block runtime).

Another problem your code did not account for is that loading 5GB JSON from disk and then parsing it requires quite a bit of RAM, which is why [`for_each_record` was used](https://github.com/near/near-indexer-for-explorer/blob/master-nearcore/src/db_adapters/genesis.rs#L63), that processes the file in a streaming manner instead of loading everything into RAM. That `for_each_record` expecting sync callback function was the main reason for `block_on` helper existence as we needed to run async code that pushes data into the database from this sync function that we wanted to run in a separate thread to avoid blocking the async runtime where `store_genesis_records` was executed (sync `main` -> async `main` -> async `store_genesis_records` -> sync `for_each_record` handler -> async `store_accounts_from_genesis`)

Reply:
Thanks for the explanation @frol, I didn't even consider how this code would behave when dealing with a large genesis file.

> I would still consider extracting it into a separate thread just to avoid unnecessary surprises

Given how long a large genesis file will take to download, and then process, I wonder if blocking would be beneficial here. If we do this in parallel with streaming/storing block data, we'd likely encounter data which is dependant on genesis which would result in continuous retries until either the data is processed by the genesis thread or the number of retries is reached. I'm not sure how likely this is, but it seems pointless to start streaming blocks if we're going to run in to these blockers anyway. Was this a problem with the `nearcore` indexer?

Regardless I think there are two things I can do now to avoid the RAM issue:
1. Stream genesis file download to disk rather loading the entire response in memory and then writing to disk
2. Stream parsing of genesis file/records (after scanning `nearcore` code I'm not sure this is possible but will continue looking in to it)

Reply:
> I wonder if blocking would be beneficial here.

Well, if so, it is better to block explicitly with `.await` instead of implicitly relying on the fact that the whole async runtime thread will get blocked here. I think we should consider this currently blocking implementation as buggy since async functions must cooperate (do not block for too long).

> If we do this in parallel with streaming/storing block data

We should not do that in parallel with streaming the block data before we finish processing genesis exactly due to the named reasons.

> Was this a problem with the nearcore indexer?

Nope. The problem with the nearcore indexer was that we wanted nearcore to be able to do its work - connect to other nodes, sync and validate blocks, serve RPC, etc. If we block async runtime with a long-running blocking operation, none of the above can happen.

> Regardless I think there are two things I can do now to avoid the RAM issue:

An ideal implementation would use async read that would be directly receiving the data from the network or reading from disk (if users downloaded the genesis file manually), and parsing that on the fly, so we won't even need to store it on disk if we can read from the network. Yet, that might be too complicated and we might even fall back on asking users to manually download the genesis file (e.g. provide `curl` / `wget` command if there is no file on disk) as downloading big genesis files would ideally report some progress...

Reply:
Had a chat with @frol offline, here's a summary:

**Problems:**
* This function is the combination of sync/async functionality as Frol mentions, specifically that `Genesis::from_file()` is sync while `store_genesis_records()` and the rest of it's implementation is async
* The genesis file is first loaded in to memory and then each record is iterated, this requires a lot of RAM with large genesis files, i.e. testnet.

**Next steps:**
1. Stream download of genesis file and show progress in logs
2. Make genesis file validation configurable via CLI flag to allow faster processing of genesis records for when we know the genesis file is valid
3. Avoid re-downloading genesis file if it already exists on disk. This also plays nicely in to the flag introduced in (2.)
4. Use `Genesis::new_with_path()` to avoid loading entire file in to memory. Use `for_each_record()` to stream genesis records. As this is synchronous, add the old blocking logic back in so that it works with the async diesel code

Reply:
@frol @khorolets I have implemented most of the steps above, but currently, am still using `Genesis::from_file()`. I attempted to use `Genesis::new_with_path()`; however, it seems that it's still not possible to prevent loading the entire file into memory.

The reason for this is because `new_with_path<P: AsRef<Path>>(config: GenesisConfig, records_file: P)` requires `GenesisConfig` as an argument and constructing it involves [reading the entire file](https://github.com/near/nearcore/blob/0c673d4470307236aa8b3a17d51822c6361eb7a3/core/chain-configs/src/genesis_config.rs#L276), exactly what we are trying to avoid.

I tried passing in `GenesisConfig::default()`, as we may be able to work around not having `config.genesis_height`, but this doesn't work as `GenesisValidationMode` is not configurable and [defaults to](https://github.com/near/nearcore/blob/0c673d4470307236aa8b3a17d51822c6361eb7a3/core/chain-configs/src/genesis_config.rs#L465) `Full`. Consequently, validation fails as the defaults are insufficient. Additionally, all records are streamed/validated, but not loaded in to memory completely, potentially causing some unwanted delays.

I'm not sure what the best path forward is here. Maybe I have misunderstood some of this. Would appreciate some guidance here from you two! :)

Reply:
Discussed with @khorolets offline. Unfortunately, it doesn't look like there is any way around the problems outlined above without some work on the `nearcore` side. We've therefore decided to:
1. Stick with loading the entire file in to memory
2. Add a note to `README.md` outlining the RAM concern

Note for (1.) I've decided to use `Genesis::new_with_path()` rather than `Genesis::from_file()`. While it does load the entire file for parsing `GenesisConfig`, it won't load all `GenesisRecord`s in to memory, which allows us to stream the records.